util/topsql/tracecpu/profile.go

```diff
@@ -0,0 +1,424 @@
+// output the SQLCPUResult slice. Want to know more information about profile labels, see https://rakyll.org/profiler-labels/
+// The sql_digest label is been set by `SetSQLLabels` function after parse the SQL.
+// The plan_digest label is been set by `SetSQLAndPlanLabels` function after build the SQL plan.
+// Since `sqlCPUProfiler` only care about the cpu time that consume by (sql_digest,plan_digest), the other sample data
+// without those label will be ignore.
+func (sp *sqlCPUProfiler) parseCPUProfileBySQLLabels(p *profile.Profile) []SQLCPUResult {
+	sqlMap := make(map[string]*sqlStats)
+	idx := len(p.SampleType) - 1
+	for _, s := range p.Sample {
+		digests, ok := s.Label[labelSQLDigest]
+		if !ok || len(digests) == 0 {
+			continue
+		}
+		for _, digest := range digests {
+			stmt, ok := sqlMap[digest]
+			if !ok {
+				stmt = &sqlStats{
+					plans: make(map[string]int64),
+					total: 0,
+				}
+				sqlMap[digest] = stmt
+			}
+			stmt.total += s.Value[idx]
+
+			plans := s.Label[labelPlanDigest]
+			for _, plan := range plans {
+				stmt.plans[plan] += s.Value[idx]
+			}
+		}
+	}
+	return sp.createSQLStats(sqlMap)
+}
+
+func (sp *sqlCPUProfiler) createSQLStats(sqlMap map[string]*sqlStats) []SQLCPUResult {
+	stats := make([]SQLCPUResult, 0, len(sqlMap))
+	for sqlDigest, stmt := range sqlMap {
+		stmt.tune()
+		for planDigest, val := range stmt.plans {
+			stats = append(stats, SQLCPUResult{
+				SQLDigest:  []byte(sqlDigest),
+				PlanDigest: []byte(planDigest),
+				CPUTimeMs:  uint32(time.Duration(val).Milliseconds()),
+			})
+		}
+	}
+	return stats
+}
+
+type sqlStats struct {
+	plans map[string]int64
+	total int64
+}
+
+// tune use to adjust sql stats. Consider following situation:
+// The `sqlStats` maybe:
+//     plans: {
+//         "table_scan": 200ms, // The cpu time of the sql that plan with `table_scan` is 200ms.
+//         "index_scan": 300ms, // The cpu time of the sql that plan with `table_scan` is 300ms.
+//       },
+//     total:      600ms,       // The total cpu time of the sql is 600ms.
+// total_time - table_scan_time - index_scan_time = 100ms, and this 100ms means those sample data only contain the
+// sql_digest label, doesn't contain the plan_digest label. This is cause by the `pprof profile` is base on sample,
+// and the plan digest can only be set after optimizer generated execution plan. So the remain 100ms means the plan
+// optimizer takes time to generated plan.
+// After this tune function, the `sqlStats` become to:
+//     plans: {
+//         ""          : 100ms,  // 600 - 200 - 300 = 100ms, indicate the optimizer generated plan time cost.
+//         "table_scan": 200ms,
+//         "index_scan": 300ms,
+//       },
+//     total:      600ms,
+func (s *sqlStats) tune() {
+	if len(s.plans) == 0 {
+		s.plans[""] = s.total
+		return
+	}
+	planTotal := int64(0)
+	for _, v := range s.plans {
+		planTotal += v
+	}
+	optimize := s.total - planTotal
+	if optimize <= 0 {
+		return
+	}
+	s.plans[""] += optimize
```

Comment:
Maybe it should be = instead of +=?

```suggestion
	s.plans[""] = optimize
```

Reply:
No, It should be `+=` instead of `=`.